module/state_synchronization/requester/execution_data_requester.go

```diff
@@ -0,0 +1,688 @@
+func New(
+	log zerolog.Logger,
+	edrMetrics module.ExecutionDataRequesterMetrics,
+	datastore datastore.Batching,
+	blobservice network.BlobService,
+	eds state_synchronization.ExecutionDataService,
+	processedHeight storage.ConsumerProgress,
+	processedNotifications storage.ConsumerProgress,
+	state protocol.State,
+	headers storage.Headers,
+	results storage.ExecutionResults,
+	cfg ExecutionDataConfig,
+) (state_synchronization.ExecutionDataRequester, error) {
+	e := &executionDataRequester{
+		log:                  log.With().Str("component", "execution_data_requester").Logger(),
+		ds:                   datastore,
+		bs:                   blobservice,
+		eds:                  eds,
+		metrics:              edrMetrics,
+		headers:              headers,
+		results:              results,
+		config:               cfg,
+		finalizationNotifier: engine.NewNotifier(),
+	}
+
+	executionDataNotifier := engine.NewNotifier()
+	rootHeight := e.config.StartBlockHeight - 1
+
+	var err error
+
+	// jobqueue Jobs object that tracks sealed blocks by height. This is used by the blockConsumer
+	// to get a sequential list of sealed blocks.
+	sealedBlockReader := jobs.NewSealedBlockReader(state, headers)
+
+	// jobqueue Consumer that listens for block finalization events, then checks if there are new
+	// sealed blocks. If there are, it starts at most `fetchWorkers` number of workers to process them.
+	// when a sealed block's execution data has been downloaded, it persists the highest consecutive
+	// downloaded height. That way, if the node crashes, it has a record of the lowest missing height,
+	// and can resume from there.
+	e.blockConsumer, err = jobqueue.NewReadyDoneAwareConsumer(
+		log.With().Str("module", "block_consumer").Logger(),
+		processedHeight,                  // read and persist the downloaded height
+		sealedBlockReader,                // read sealed blocks by height
+		e.processBlockJob,                // process the sealed block job to download its execution data
+		e.finalizationNotifier.Channel(), // to listen to finalization events to find newly sealed blocks
+		rootHeight,                       // initial "last processed" height for empty db
+		fetchWorkers,                     // the number of concurrent workers
+		e.config.MaxSearchAhead,          // max number of unsent notifications to allow before pausing new fetches
+		// notifies notificationConsumer when new ExecutionData blobs are available
+		func(module.JobID) { executionDataNotifier.Notify() },
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create block consumer: %w", err)
+	}
+
+	// jobqueue Jobs object tracks downloaded execution data by height. This is used by the
+	// notificationConsumer to get downloaded execution data. It also  maintains a cache of the
+	// downloaded execution data to reduce disk reads.
+	e.executionDataCache = jobs.NewExecutionDataReader(
+		e.config.MaxCachedEntries,          // max number of ExecutionData objects to cache
```

Comment:
I think it's fine that there could be a gap in the cached entries in the situation where a node isn't keeping up.

The cache is mostly there as an optimization for the happy path to reduce the amount of disk IO used by operators, particularly ones running in cloud environment where they pay per unit of IO. It would also help reduce latency for blocks with large ExecutionData, but with SSDs, that probably a marginal difference.

ExecutionData objects on average are ~40KB, but could spike up to 10s or 100s of MB depending on what transactions were in the block. I picked the default cache size to limit the max memory to something reasonable given we recommend 16 GB of memory for ANs. It would probably be fine in the 100-200 range as well.

I'm ok increasing the cache size, but since there's isn't any cache data size limits in place, I'm reluctant to make the default too high.